### PR TITLE
AWS Messaging Framework Design

### PR DESCRIPTION
**Pull request by ashovlin**
_Context: The AWS .NET team is exploring creating an AWS native framework that simplifies development of .NET message processing applications using AWS services. This PR is to collect feedback on the design document. You can also read the [rendered view](https://github.com/aws/dotnet/blob/a9efc5189e111101c566c86aed30be157b6bde41/designs/messaging/messaging-framework.md)._

The purpose of the framework would be to reduce the amount of boiler-plate code developers need to write. The primary responsibilities of the proposed framework are: 

- **Handling the message routing** - In a publisher, the framework will handle routing the messages to the correct queue/topic/eventbus. In a consumer process, it will route the particular message type to the appropriate business logic.
- **Handling the overall message lifecycle**  - The framework will handle serializing/deserializing the message to .NET objects, keeping track of the message visibility while it is being processed, and deleting the message when completed.

Here is an example showing a sample publisher and handler for a hypothetical `OrderInfo` message.

Sample publisher:
```csharp
[ApiController]
[Route("[controller]")]
public class OrderController : ControllerBase
{
    // See later in the design for how this was configured and mapped to the queue
    private readonly IMessagePublisher _publisher;

    public OrderController(IMessagePublisher publisher)
    {
        _publisher = publisher;
    }

    [HttpPost]
    public async Task Post([FromBody] OrderInfo orderInfo)
    {
        // Add internal metadata to the OrderInfo object 
        // we received, or any other business logic
        orderInfo.OrderTime = DateTime.UtcNow;
        orderInfo.OrderStatus = OrderStatus.Recieved;

        // The updated OrderInfo object will also be serialized as the SQS message
        await _publisher.PublishAsync(orderInfo);
    }
}
```

Sample handler:
```csharp
// See later in the design for how this was configured and mapped to the queue
public class OrderInfoHandler : IMessageHandler<OrderInfo>
{
    public async Task<MessageStatus> HandleAsync(MessageEnvelope<OrderReceived> message, CancellationToken cancellationToken = default(CancellationToken))
    {
        // Here we're reading from the message within the metadata envelope
        var productId = message.Message.ProductId;
        
        // Here we can do our business logic based on what is in the message
        await UpdateInventory(productId);
        await PrintShippingLabel(productId, message.Message.CustomerId);

        // Indicate that OrderInfo has been processed successfully
        return MessageStatus.Success;
    }
}
```

On either this PR or on specific section(s) of the design, please comment with:
* What are your thoughts and feedback on the proposed framework design?
* What are your thoughts on the initial MVP scope?
* Which AWS messaging services do you use?
    * Amazon SQS 
    * Amazon SNS
    * Amazon EventBridge
    * Amazon Kinesis
    * Amazon MQ
    * Other (specify which ones)
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: designs/messaging/messaging-framework.md [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+        builder.AddPublishingQueue<OrderInfo>(awsResourceIds.OrderInfoQueue, "OrderMessageType");
+        
+        builder.AddPublishingTopic<HealthStatus>(awsResourceIds.HealthStatusTopic);
+        
+        // One could also specify PublishMetadata that will be set in the 
+        // message envelope for ALL messages published by this application
+        builder.AddPublishingEventBus<InventoryCheck>(awsResourceIds.WarehouseEventBus, new PublishMetadata
+        {
+            MessageType = "InventoryCheckMessage",
+            Source = "OrderHandlerSystem"
```
Comment (TobbenTM) [inline designs/messaging/messaging-framework.md:123]: Would it also be possible to set metadata on the message bus as a whole? I can imagine that the above example of `Source` would naturally apply to all publishers in this assembly?

File: designs/messaging/messaging-framework.md [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+```csharp
+public IConfiguration Configuration { get; }
+
+public void ConfigureServices(IServiceCollection services)
+{
+    // Pull out physical AWS resource ids from .NET Configuration framework
+    var awsResourceIds = new AWSResourceIds();
+    Configuration.GetSection(AWSResourceIds.SectionName).Bind(awsResourceIds);
+            
+    services.AddAWSMessageBus(builder =>
```
Comment (TobbenTM) [inline designs/messaging/messaging-framework.md:107]: Have you considered how to deal with serialization options? Our events might have polymorphic properties on them requiring a suitable STJ converter, there should be a way to either tell this library to use an already registered serializer, or pass in a `System.Text.Json.JsonSerializerOptions`.

**Comment by ashovlin** [comment]
Thank you for the feedback so far! We've just pushed a revision to the design document with:
* Expanded the community library section at the top to clarify the niche we see this framework filling.
* Multiple commenters suggested CloudEvents, which admittedly wasn't on our radar initially. The JSON Event Format for CloudEvents is very similar to the envelope schema we had originally proposed, so we're now exploring using the CloudEvents format.

**Comment by jeffroxpanse** [comment]
Nice document, and timely -- I'm (re)creating this wheel right now.

As this relates to message lifecycle, are there forthcoming thoughts to operations?

- Distributed subscription references in a micro-service architecture?
- How will queues, topics, subscriptions generated/maintained?
- Are there expected/required security policies needed?
- Thoughts of support for highly-durable operations? (cross-region, for example)

I'm definitely following this project!

**Comment by ashovlin** [comment]
We've moved the design doc over to the new repository at https://github.com/awslabs/aws-dotnet-messaging/blob/main/docs/design/message-processing-framework-design.md.

Feel free to follow along with the development and leave comments/issues over on https://github.com/awslabs/aws-dotnet-messaging. Thanks again for the feedback here!

**Pull request closed without merging**